crates/holochain/src/conductor/interface/websocket.rs

```diff
@@ -165,58 +203,111 @@ async fn recv_incoming_admin_msgs<A: InterfaceApi>(api: A, rx_from_iface: Websoc
-                InterfaceResult::Ok(())
-            }
-            .await
-            {
-                error!(?err, "error emitting signal");
+        loop {
+            match rx_from_cell.recv().await {
+                // We missed some signals, but the channel is still open
+                Err(tokio::sync::broadcast::error::RecvError::Lagged(_)) => {
+                    warn!("SignalChannelOverloaded");
+                    continue;
+                }
+                Ok(item) => return Some((item, rx_from_cell)),
+                _ => {
+                    warn!("SignalChannelClosed");
+                    return None;
+                }
             }
         }
-    }));
+    });
 
-    tokio::task::spawn(rx_from_iface.for_each_concurrent(4096, move |msg| {
-        let api = api.clone();
-        async move {
-            if let Err(err) = handle_incoming_message(msg, api).await {
-                error!(?err, "error handling websocket message");
+    task_list
+        .0
+        .push(tokio::task::spawn(rx_from_cell.for_each_concurrent(
+            CONCURRENCY_COUNT,
+            move |signal| {
+                let tx_to_iface = tx_to_iface.clone();
+                async move {
+                    trace!(msg = "Sending signal!", ?signal);
+                    if let Err(err) = async move {
+                        tx_to_iface.signal(signal).await?;
+                        InterfaceResult::Ok(())
+                    }
+                    .await
+                    {
+                        error!(?err, "error emitting signal");
+                    }
+                }
+            },
+        )));
+
+    let rx_from_iface =
+        futures::stream::unfold(rx_from_iface, move |mut rx_from_iface| async move {
+            match rx_from_iface.recv().await {
+                Ok(r) => Some((r, rx_from_iface)),
+                Err(err) => {
+                    info!(?err);
+                    None
+                }
             }
-        }
-    }));
+        });
+
+    task_list
+        .0
+        .push(tokio::task::spawn(rx_from_iface.for_each_concurrent(
+            CONCURRENCY_COUNT,
+            move |msg| {
+                let api = api.clone();
+                async move {
+                    if let Err(err) = handle_incoming_message(msg, api).await {
+                        error!(?err, "error handling websocket message");
+                    }
+                }
+            },
+        )));
 }
 
 /// Handles messages on all interfaces
-async fn handle_incoming_message<A>(ws_msg: WebsocketMessage, api: A) -> InterfaceResult<()>
+async fn handle_incoming_message<A>(
+    ws_msg: ReceiveMessage<A::ApiRequest>,
+    api: A,
+) -> InterfaceResult<()>
 where
     A: InterfaceApi,
 {
-    let (bytes, respond) = ws_msg;
-    Ok(respond
-        .respond(api.handle_request(bytes.try_into()).await?.try_into()?)
-        .await?)
+    match ws_msg {
+        ReceiveMessage::Signal(_) => {
```

Comment:
Is there a reason this is even an option? It's hard to tell here but this looks like one of our types so maybe we can just remove this variant?

Reply:
The protocol is symmetric. We need the signals for going out, so we have to be able to deal with the potential for them coming in. This is how it was already written, I didn't want to introduce any breaking protocol changes in this PR.

Reply:
Ahh of course, yes that makes sense